### PR TITLE
extend prepare_datasets logic to allow different data splitting methods

### PR DESCRIPTION
**Pull request by bas0N**

### CODE REVIEW

File: notebooks/data_preparation_and_analysis.ipynb [hunk 1]
```diff
@@ -1,226 +1,319 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "c:\\Users\\wojciech.basinski\\iml\\venv\\Lib\\site-packages\\IPython\\core\\magics\\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
```
Comment (mytkom) [inline notebooks/data_preparation_and_analysis.ipynb]: Clear notebook outputs, please.

File: notebooks/data_preparation_and_analysis.ipynb [hunk 2]
```diff
@@ -1,226 +1,319 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "balanced_user_files = balance_recordings_by_user(wav_files, max_files_per_user=3)\n",
+    "print(\"\\nFinal Balanced User Files:\", balanced_user_files)\n",
+    "\n",
+    "# Balance recordings per device\n",
+    "balanced_device_files = balance_recordings_by_device(wav_files, max_files_per_device=2)\n",
+    "print(\"\\nFinal Balanced Device Files:\", balanced_device_files)"
```
Comment (mytkom) [inline notebooks/data_preparation_and_analysis.ipynb]: The purpose of balancing the data is to improve Neural Network training results. I would like you to use balanced data for Neural Network training and let us know about any improvements. Could you do it as part of this PR?

File: src/data_processing.py [hunk 3]
```diff
@@ -1,231 +1,348 @@
...
+    filename = os.path.basename(filepath)
+    parts = filename.split('_')
+    if len(parts) >= 3:
+        user = parts[0]
+        script = parts[1]
+        device = parts[2].split('.')[0]  # Removing the extension
+        print(f"Extracted Metadata - User: {user}, Script: {script}, Device: {device}")
+        return user, script, device
+    print(f"Failed to extract metadata from: {filepath}")
+    return None, None, None
```
Comment (mytkom) [inline src/data_processing.py]: That is a handy function; I like it!

File: notebooks/data_preparation_and_analysis.ipynb [hunk 4]
```diff
@@ -1,226 +1,319 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "soa_train_full_clips = SOAAudioClips(train_files)\n",
+    "soa_test_full_clips = SOAAudioClips(test_files)\n",
+    "soa_val_full_clips = SOAAudioClips(val_files)\n",
```
Comment (mytkom) [inline notebooks/data_preparation_and_analysis.ipynb]: I want `train_files` to be balanced to check how it improves the training process. And I think that it should be balanced in both ways user + device, not only one of them at the time. Also, the authorized/unauthorized ratio is to be 1:1 in the training dataset. 
example:
Our dataset:
```
speaker: f1, recordings 1-8, devices {A, B}
speaker: m1, recordings 1-8, devices {A, B}
speaker: m2, recordings 1-8, devices {A, B}
```
Authorized: `{ f1 }`
Unauthorized: `{ m1, m2 }`
I think that a reasonable balance of every aspect of this dataset would be to:
1. m1, m2 should have 2 times fewer recordings than f1 because the unauthorized set size is 2 times greater.
2. When choosing a recording to be in train dataset, we should include this recording from every available device. We don't want common recordings in train/validation/test datasets to prevent testing on training data.

According to this I would split the example dataset into for example:
Train: `(f1, 1, A), (f1, 1, B),(f1, 2, A), (f1, 2, B),(f1, 3, A), (f1, 3, B),(f1, 4, A), (f1, 4, B), (m1, 1, A), (m1, 1, B), (m1, 2, A), (m1, 2, B), (m2, 1, A), (m2, 1, B), (m2, 2, A), (m2, 2, B)`
Validation and Test - I would try to make the same distribution of recordings as in the input dataset (we want to test a real-world scenario) but from the files not included in `train_files`. And remember, about point 2, we don't want to ex. (m1,7,A) in test and (m1,7,B) in validation because these are recordings of the same thing, just different noises.

What do you think? Does this make sense to you?

Comment (bas0N) [reply notebooks/data_preparation_and_analysis.ipynb]: I have managed to split the data (from all catalogs (except cleanraw, clean, production) in the following way:

`Train Dataset Statistics: {'Total Files': 378, 'Authorized': 194, 'Unauthorized': 184, 'Speakers': Counter({'m8': 35, 'f8': 34, 'm6': 32, 'm3': 31, 'f1': 31, 'f7': 31, 'f6': 19, 'm1': 17, 'f5': 17, 'm10': 14, 'f9': 14, 'm7': 14, 'm9': 14, 'f4': 13, 'f3': 13, 'f10': 13, 'm2': 11, 'm5': 11, 'm4': 10, 'f2': 4}), 'Devices': Counter({'ipad': 243, 'iphone': 135}), 'Rooms': Counter({'balcony1': 88, 'livingroom1': 81, 'bedroom1': 80, 'confroom2': 51, 'office1': 43, 'confroom1': 35})}

Validation Dataset Statistics: {'Total Files': 10, 'Authorized': 5, 'Unauthorized': 5, 'Speakers': Counter({'f1': 2, 'f2': 1, 'm6': 1, 'm2': 1, 'f10': 1, 'f7': 1, 'm10': 1, 'm3': 1, 'f3': 1}), 'Devices': Counter({'ipad': 6, 'iphone': 4}), 'Rooms': Counter({'bedroom1': 4, 'balcony1': 2, 'livingroom1': 2, 'confroom2': 1, 'office1': 1})}

Test Dataset Statistics: {'Total Files': 10, 'Authorized': 6, 'Unauthorized': 4, 'Speakers': Counter({'f8': 3, 'f1': 2, 'f3': 2, 'f2': 1, 'm8': 1, 'f5': 1}), 'Devices': Counter({'ipad': 8, 'iphone': 2}), 'Rooms': Counter({'confroom2': 3, 'balcony1': 2, 'bedroom1': 2, 'livingroom1': 1, 'office1': 1, 'confroom1': 1})}
`

Compared to our regular split from example folder like ipad_balcony1

`
Train Dataset Statistics: {'Total Files': 64, 'Authorized': 19, 'Unauthorized': 45, 'Speakers': Counter({'f5': 5, 'm1': 4, 'f3': 4, 'f1': 4, 'm10': 4, 'm9': 4, 'm7': 4, 'f9': 4, 'f8': 4, 'm2': 4, 'm6': 3, 'm4': 3, 'm3': 3, 'm8': 3, 'f7': 2, 'm5': 2, 'f4': 2, 'f2': 2, 'f6': 2, 'f10': 1}), 'Devices': Counter({'ipad': 64}), 'Rooms': Counter({'balcony1': 64})}

Validation Dataset Statistics: {'Total Files': 16, 'Authorized': 6, 'Unauthorized': 10, 'Speakers': Counter({'f4': 2, 'f10': 2, 'f7': 2, 'm3': 2, 'm1': 1, 'f1': 1, 'm2': 1, 'f2': 1, 'm8': 1, 'm7': 1, 'm9': 1, 'm5': 1}), 'Devices': Counter({'ipad': 16}), 'Rooms': Counter({'balcony1': 16})}

Test Dataset Statistics: {'Total Files': 20, 'Authorized': 5, 'Unauthorized': 15, 'Speakers': Counter({'f6': 3, 'm6': 2, 'm4': 2, 'f2': 2, 'f10': 2, 'm5': 2, 'm10': 1, 'f9': 1, 'f8': 1, 'f7': 1, 'f4': 1, 'f3': 1, 'm8': 1}), 'Devices': Counter({'ipad': 20}), 'Rooms': Counter({'balcony1': 20})}`

gives a decent difference :)

File: notebooks/data_preparation_and_analysis.ipynb (hunk 2 above)
Comment (bas0N) [reply notebooks/data_preparation_and_analysis.ipynb]: My laptop is dying during the training part, I need to come up with some online way to do this

**Comment by mytkom** [comment]
Nice job! I like the separate notebook and how it is designed now.

Nevertheless, I see a significant issue: the split ratio for train, validation and test datasets is broken for the result split. You specify the training dataset as 70% of files, but after excluding some files, it is 95% (because of this min(10, len(train)) thing in case of empty val/test set). Maybe you can inverse the logic to first split the dataset into small locally balanced subsets and then distribute subsets according to training/validation/test ratio. An additional minor issue is that (if I am not mistaken) we do not check that every speaker is present in the training dataset and I think we should do this.

**Comment by bas0N** [comment]
I have reached the following split:
--- Training Set Statistics ---
Total Samples: 432
Total Speakers: 20
Authorized Samples: 216
Unauthorized Samples: 216
Authorized to Unauthorized Ratio: 216:216

Samples per Speaker:
  f1: 36
  f10: 17
  f2: 17
  f3: 18
  f4: 21
  f5: 17
  f6: 14
  f7: 36
  f8: 36
  f9: 11
  m1: 15
  m10: 17
  m2: 15
  m3: 36
  m4: 11
  m5: 15
  m6: 36
  m7: 15
  m8: 36
  m9: 13

--- Validation Set Statistics ---
Total Samples: 240
Total Speakers: 20
Authorized Samples: 72
Unauthorized Samples: 168
Authorized to Unauthorized Ratio: 72:168

Samples per Speaker:
  f1: 12
  f10: 12
  f2: 12
  f3: 12
  f4: 12
  f5: 12
  f6: 12
  f7: 12
  f8: 12
  f9: 12
  m1: 12
  m10: 12
  m2: 12
  m3: 12
  m4: 12
  m5: 12
  m6: 12
  m7: 12
  m8: 12
  m9: 12

--- Test Set Statistics ---
Total Samples: 240
Total Speakers: 20
Authorized Samples: 72
Unauthorized Samples: 168
Authorized to Unauthorized Ratio: 72:168

Samples per Speaker:
  f1: 12
  f10: 12
  f2: 12
  f3: 12
  f4: 12
  f5: 12
  f6: 12
  f7: 12
  f8: 12
  f9: 12
  m1: 12
  m10: 12
  m2: 12
  m3: 12
  m4: 12
  m5: 12
  m6: 12
  m7: 12
  m8: 12
  m9: 12

I think it can be improved to 70/15/15 train/val/test ratio, yet it has very important features:

- All speakers are represented in the training set.
- Training set has a 1:1 ratio of authorized to unauthorized samples.
- No speaker-script combination is present in more than one set. (no testing on the data it was trained)

The results of model training with it can be seen here: 
1. OriginalSizeCNN: https://api.wandb.ai/links/wch-basinski-politechnika-warszawska/6q32y095
2. TutorialCNN: https://api.wandb.ai/links/wch-basinski-politechnika-warszawska/hs0lfkug
3. TutorialCNN without standardization: https://api.wandb.ai/links/wch-basinski-politechnika-warszawska/rr915m74
